Refactor coverage instrumentation

peterschrammel opened the pull request:
Splits cover.cpp into several files and class-ifies various instrumenters and filters applied

Companions:
* https://github.com/diffblue/test-gen/pull/1063
* https://github.com/diffblue/symex/pull/2

subsumes https://github.com/diffblue/cbmc/pull/1348
depends on: https://github.com/diffblue/cbmc/pull/1685 (first commit of this PR)

smowton commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -0,0 +1,237 @@
+
+void cover_basic_blockst::output(std::ostream &out) const
+{
+  for(block_mapt::const_iterator
```
Convert to for-each loop while we're here?

smowton commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -0,0 +1,237 @@
+  INVARIANT(!cover_set.empty(),
+            "covered lines set must not be empty");
+  std::vector<unsigned>
+    line_list{cover_set.begin(), cover_set.end()};
```
@reuk will no doubt know better, but is there any advantage to using init-list-style construction here (briefly suggesting a list of iterators) rather than constructor notation (`line_list(cover_set.begin(), cover_set.end())`)?

reuk replied:
Basically no difference, but I prefer the curly brace syntax as it allows constructor calls to be distinguished from free function calls.

smowton replied:
Ahh, I always expect something behaving like aggregate init (e.g. `std::vector<int> {1,2,3}`) when I see braced init. Matter of taste I suppose, but we should arrive at some convention for the codebase and exercise it consistently.

smowton commented on src/goto-instrument/cover_filter.cpp:
```diff
@@ -0,0 +1,228 @@
+    id2string(identifier), string_matcher, regex_matcher);
+}
+
+/// Call a goto_program non-trivial if it has: * Any declarations * At least 2
```
Line-break bulleted list

smowton commented on src/goto-instrument/cover_instrument_branch.cpp:
```diff
@@ -0,0 +1,62 @@
+  goto_programt::targett &i_it,
+  const cover_basic_blockst &basic_blocks) const
+{
+  if(i_it->is_assert())
```
In the case where several instrumenters are being run, this will destroy the previous one's work. These may be redundant in the face of assert-to-assume always preceding instrumentation, or if this is needed for other reasons then we should simply rule out multiple coverage types for now.

peterschrammel replied:
Well spotted.

smowton commented on src/goto-instrument/cover_instrument_branch.cpp:
```diff
@@ -0,0 +1,62 @@
+    i_it->source_location=source_location;
+    initialize_source_location(i_it, false_comment, function);
+
+    i_it++;
```
Suggest std::advance(i_it, 2) to make clear this isn't a typo

smowton reviewed (changes requested):
One significant problem to look at and a few cleanups that might be worth getting in while we're touching every line

tautschnig commented:
@peterschrammel Requires a rebase.

peterschrammel committed:
Replace cover-function-only by cover-include-pattern

cover-function-only (instrumentating coverage goals in the
entry point function only) is now implemented with the help
of cover-include-pattern (regex matching of functions to
be instrumented). This is to avoid potential consistencies
due to providing two mechanisms for filtering functions.
cover-function-only has priority over the given
cover-include-pattern.

peterschrammel committed:
Split cover into several files

peterschrammel committed:
Introduce cover instrumenter

peterschrammel committed:
Document cover functions

peterschrammel committed:
Factorize source location initialization

martin-cs commented on src/goto-instrument/cover.cpp:
```diff
@@ -11,1687 +11,196 @@ Date: May 2016
+/// \return a coverage criterion or throws an exception
+coverage_criteriont
+parse_coverage_criterion(const std::string &criterion_string)
+{
```
Is this a violation of the "command line parsing should be done in the binary / front-end directories" principle?

peterschrammel replied:
Command line parsing is a bit involved here. Repeating the code for each binary cannot be the solution. I've provided utility functions here that should be used by the binary to avoid duplication.

martin-cs commented on src/goto-instrument/cover.cpp:
```diff
@@ -11,1687 +11,196 @@ Date: May 2016
-    }
+    std::stringstream s;
+    s << "unknown coverage criterion " << '\'' << criterion_string << '\'';
+    throw s.str();
```
This feels a bit wrong as I'd want a user error to give a meaningful error message, not throw an exception.

martin-cs commented on src/goto-instrument/cover.cpp:
```diff
@@ -1718,8 +227,23 @@ bool instrument_cover_goals(
+  std::string cover_include_pattern =
+    cmdline.get_value("cover-include-pattern");
+  if(cmdline.isset("cover-function-only"))
+  {
```
I think command line parsing should be done in the binary directories.

martin-cs commented on src/goto-instrument/cover.h:
```diff
@@ -15,68 +15,34 @@ Date: May 2016
+  PATH,
+  MCDC,
+  ASSERTION,
+  COVER
```
This option is far from obvious.  Isn't this enum to saw what to cover?

martin-cs commented on src/goto-instrument/cover.h:
```diff
@@ -15,68 +15,34 @@ Date: May 2016
-  coverage_goalst &,
-  bool function_only=false,
-  bool ignore_trivial=false,
-  const std::string &cover_inclue_pattern="");
```
A definite improvement.

martin-cs commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -0,0 +1,227 @@
+
+Module: Coverage Instrumentation
+
+Author: Daniel Kroening
```
If this is a new file, shouldn't it say Peter Schrammel?

martin-cs commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -0,0 +1,227 @@
+      it->is_goto() || it->is_function_call() || it->is_assume();
+#else
+      it->is_goto() || it->is_function_call();
+#endif
```
I guess assume is the issue?  Can this be resolved rather than adding #if 0?

peterschrammel replied:
I've left this as is since it was already commented out. Possibly a flag is needed to have some control over what is considered a basic block.

martin-cs commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -0,0 +1,227 @@
+  block_mapt::const_iterator it = block_map.find(t);
+  INVARIANT(it != block_map.end(), "instruction must be part of a block");
+  return it->second;
+}
```
I'm kind of surprised we don't have a utility function for 'find basic blocks', similarly to finding loops.  It would be good if this could be split out to be something general.

peterschrammel replied:
#1686

martin-cs commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -0,0 +1,227 @@
+    if(block_info.representative_inst == goto_program.instructions.end())
+    {
+      if(!it->source_location.get_java_bytecode_index().empty())
+      {
```
This seems very language specific; what happens if the input is not JVM code?  Would it be better to separate basic blocks and JVM instructions as different coverage options?

peterschrammel replied:
Yes, that's a good idea. Noted for a future refactoring: https://github.com/diffblue/cbmc/issues/1686

martin-cs commented on src/goto-instrument/cover_filter.cpp:
```diff
@@ -0,0 +1,233 @@
+
+Module: Coverage Instrumentation
+
+Author: Daniel Kroening
```
Again; cut and paste issue.

martin-cs commented on src/goto-instrument/cover_filter.cpp:
```diff
@@ -0,0 +1,233 @@
+///  * Any declarations
+///  * At least 2 branches
+///  * At least 5 assignments
+/// \param identifier: a function name
```
Is this just an arbitrary choice or is there some deeper significance?

martin-cs commented on src/goto-instrument/cover_filter.cpp:
```diff
@@ -0,0 +1,233 @@
+  const std::string &filename,
+  const irep_idt &mode)
+  : goal_filter_baset(message_handler)
+{
```
This seems pretty involved for a constructor.

martin-cs commented on src/goto-instrument/cover_filter.cpp:
```diff
@@ -0,0 +1,233 @@
+  if(parse_json(filename, message_handler, json))
+  {
+    throw filename + " file is not a valid json file";
+  }
```
Would it be better to write the reasons to error()?

martin-cs commented on src/goto-instrument/cover_instrument_decision.cpp:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    for(std::size_t i = 0; i < decisions.size() * 2; i++)
+      i_it++;
```
What's this for?

martin-cs reviewed (commented):
A few concerns about language specific things and a few places where it could perhaps be cleaner but over-all, pretty good.

peterschrammel committed:
Pass command line options via optionst

peterschrammel commented:
@smowton, @martin-cs, ready for re-review. I've rebased it on https://github.com/diffblue/cbmc/pull/1685.

smowton commented on src/goto-instrument/cover_instrument_condition.cpp:
```diff
@@ -0,0 +1,54 @@
+{
+  if(
+    i_it->is_assert() &&
+    i_it->source_location.get_property_class() != property_class)
```
At first reading I thought this would consider "location", "mcdc", and so on different classes, but it doesn't. Therefore suggest factoring this out into a function named `is_non_coverage_assertion` or something, and explain there that all coverage assertions are of the same class.

peterschrammel replied:
Factored out. I also don't think that there are many use cases where replacing assert by skip makes sense.

smowton reviewed (approved):
Looks good, could improve clarity by factoring the is-non-coverage-assertion test.

Relatedly, is there ever a situation where assertion -> skip is a sensible thing to do? It seems like it would always lead to code being accessible with its preconditions unchecked leading to crazy behaviour, hence our always running assert-to-assume in the product. Should we perhaps issue a warning if there are somehow still live non-coverage assertions present? If I were writing this from scratch I'd recommend an invariant that there are no non-coverage assertions, but I guess that might break some existing user?

chrisr-diffblue reviewed (approved):
Looks good to me from my understanding of the code.
